### PR TITLE
Go 1.20.4, remove some tls.Config fields, update circl

### PR DESCRIPTION
**Pull request by Lekensteyn**
High-level changes:
- Rebase to Go 1.20.4 from Go 1.20.1 (no conflicts after fixing circl vendor).
- Update circl to v1.3.3 which downgrades some go.mod updated that resulted in merge conflicts.
- Removed `tls.Config#CFControl` field (#146)
- Replaced `tls.Config#CFEventHandler` field by a context value accessible by a public `tls.CFEventHandlerContextKey` context key (#146).

**Please review** whether the last `CFEventHandler` replacement is reasonable. Should the context value be an interface type instead, or is the current `func(CFEvent)` type good enough?

The range-diff is somewhat large (24.9k lines) due to the vendor changes, but if you ignore that it should look pretty reasonable
```bash
git range-diff go1.20.1..902e08e579 go1.20.4..8284e397f7 # go1.20.1..cf go1.20.4..pwu/cf-1.20.4
```

Minor changes:
- Fixed formatting with `git rebase --exec "cd src && go fmt ./crypto/tls ./crypto/x509" go1.20.1`. Verified `git diff` and `git commit --amend -a --no-edit && git rebase --continue` after each update.
- After changes to `go.mod` or the vendor (e.g. new modules for DCs, ECH), revendored with: `(cd src && go mod vendor)`

Some remaining changes and considerations before merging:
- [x] The `CFEventHandler` removal was done in a single commit (2f61bdf8df) for ease of review, but I plan to edit the original commits (`crypto/tls: add Config.CFEventHandler to expose intra-handshake timing`, `crypto/tls: implement draft-ietf-tls-esni-13`, `Add hybrid post-quantum key agreement.`)
- [x] Drop commit `crypto/tls: add CFControl parameter to Config` and its following revert commit.
- ~~Should we also add 0x0030 for X25519Kyber768Draft00 as [suggested here](https://www.ietf.org/archive/id/draft-westerbaan-cfrg-hpke-xyber768d00-02.html#name-iana-considerations-13), or wait until it is actually allocated? @bwesterb~~ CONCLUSION: no, this can be done in a future change.

### CODE REVIEW

File: src/crypto/tls/ech.go
Comment (cjpatton) [inline src/crypto/tls/ech.go]: nit: I learned long ago that you're supposed to put stdlib imports and imports from external packages in different stanzas.
```suggestion
	"io"

        "github.com/cloudflare/circl/hpke"
	"golang.org/x/crypto/cryptobyte"
```

**Review by cjpatton — APPROVED** [review]
This looks totally reasonable to me 👍

For what it's worth, `CFControl` and `CFEventHandler` were added before `context.Context` was added to `crypto/tls`. I believe this happened around version `1.19`?

As for adding additional Xyber combos/codepoints: It may be a good idea to wait until the rebase is done, merge a PR for these changes, then fix up the patches once more.

**Review by bwesterb — APPROVED** [review]

File: src/crypto/tls/ech.go
Comment (Lekensteyn) [reply src/crypto/tls/ech.go]: Thanks for flagging, I'll fix it with the next push. I ran `gofmt` and assumed it would fix the import ordering. Apparently not good enough, so I ran `goimports` on every commit to fix it:

```
git rebase -i 94b33ada97~ --exec 'cd src && $HOME/go/bin/goimports -l -w crypto/tls/*.go crypto/x509/*.go'
git diff    # after reviewing and potential manual edits:
git commit --amend -a --no-edit && git rebase --continue
```

I also checked whether I missed anything:
```
git grep -ne \"github.com/cloudflare/ -C3 src/crypto
```

The changes compared to the initial version should be pretty non-controversial:
<details>
<summary>git diff -U2 2f61bdf8df 5e7998ded1</summary>

```diff
diff --git a/src/crypto/tls/auth.go b/src/crypto/tls/auth.go
index f1ff78e322..604e431540 100644
--- a/src/crypto/tls/auth.go
+++ b/src/crypto/tls/auth.go
@@ -6,7 +6,4 @@ package tls
 
 import (
-	circlPki "github.com/cloudflare/circl/pki"
-	circlSign "github.com/cloudflare/circl/sign"
-
 	"bytes"
 	"crypto"
@@ -19,4 +16,7 @@ import (
 	"hash"
 	"io"
+
+	circlPki "github.com/cloudflare/circl/pki"
+	circlSign "github.com/cloudflare/circl/sign"
 )
 
diff --git a/src/crypto/tls/ech.go b/src/crypto/tls/ech.go
index 0c717c0008..8e71a3a8b8 100644
--- a/src/crypto/tls/ech.go
+++ b/src/crypto/tls/ech.go
@@ -7,7 +7,8 @@ import (
 	"errors"
 	"fmt"
-	"github.com/cloudflare/circl/hpke"
 	"io"
 
+	"github.com/cloudflare/circl/hpke"
+
 	"golang.org/x/crypto/cryptobyte"
 )
diff --git a/src/crypto/tls/ech_config.go b/src/crypto/tls/ech_config.go
index ec9985121f..62de3cec54 100644
--- a/src/crypto/tls/ech_config.go
+++ b/src/crypto/tls/ech_config.go
@@ -7,7 +7,8 @@ import (
 	"errors"
 	"fmt"
+	"io"
+
 	"github.com/cloudflare/circl/hpke"
 	"github.com/cloudflare/circl/kem"
-	"io"
 
 	"golang.org/x/crypto/cryptobyte"
diff --git a/src/crypto/tls/ech_provider.go b/src/crypto/tls/ech_provider.go
index 7868a812b0..d114efb276 100644
--- a/src/crypto/tls/ech_provider.go
+++ b/src/crypto/tls/ech_provider.go
@@ -7,4 +7,5 @@ import (
 	"errors"
 	"fmt"
+
 	"github.com/cloudflare/circl/hpke"
 	"github.com/cloudflare/circl/kem"
diff --git a/src/crypto/tls/generate_cert.go b/src/crypto/tls/generate_cert.go
index 237292390c..360ce8fa80 100644
--- a/src/crypto/tls/generate_cert.go
+++ b/src/crypto/tls/generate_cert.go
@@ -20,6 +20,4 @@ import (
 	"encoding/pem"
 	"flag"
-	circlSign "github.com/cloudflare/circl/sign"
-	circlSchemes "github.com/cloudflare/circl/sign/schemes"
 	"log"
 	"math/big"
@@ -28,4 +26,7 @@ import (
 	"strings"
 	"time"
+
+	circlSign "github.com/cloudflare/circl/sign"
+	circlSchemes "github.com/cloudflare/circl/sign/schemes"
 )
 
diff --git a/src/crypto/tls/generate_delegated_credential.go b/src/crypto/tls/generate_delegated_credential.go
index c93c321a02..1b72b414d5 100644
--- a/src/crypto/tls/generate_delegated_credential.go
+++ b/src/crypto/tls/generate_delegated_credential.go
@@ -24,9 +24,10 @@ import (
 	"flag"
 	"fmt"
-	circlSign "github.com/cloudflare/circl/sign"
 	"log"
 	"os"
 	"path/filepath"
 	"time"
+
+	circlSign "github.com/cloudflare/circl/sign"
 )
 
diff --git a/src/crypto/tls/handshake_server.go b/src/crypto/tls/handshake_server.go
index dcfcfb2564..3c1d219e7a 100644
--- a/src/crypto/tls/handshake_server.go
+++ b/src/crypto/tls/handshake_server.go
@@ -6,6 +6,4 @@ package tls
 
 import (
-	circlSign "github.com/cloudflare/circl/sign"
-
 	"context"
 	"crypto"
@@ -20,4 +18,6 @@ import (
 	"io"
 	"time"
+
+	circlSign "github.com/cloudflare/circl/sign"
 )
 
diff --git a/src/crypto/tls/hpke.go b/src/crypto/tls/hpke.go
index 1361d58a0b..b3861d0951 100644
--- a/src/crypto/tls/hpke.go
+++ b/src/crypto/tls/hpke.go
@@ -7,4 +7,5 @@ import (
 	"errors"
 	"fmt"
+
 	"github.com/cloudflare/circl/hpke"
 )
diff --git a/src/crypto/tls/tls.go b/src/crypto/tls/tls.go
index aaceb4978b..973aac4379 100644
--- a/src/crypto/tls/tls.go
+++ b/src/crypto/tls/tls.go
@@ -54,6 +54,4 @@ package tls
 
 import (
-	circlSign "github.com/cloudflare/circl/sign"
-
 	"bytes"
 	"context"
@@ -69,4 +67,6 @@ import (
 	"os"
 	"strings"
+
+	circlSign "github.com/cloudflare/circl/sign"
 )
 
diff --git a/src/crypto/tls/tls_cf.go b/src/crypto/tls/tls_cf.go
index 65db04e02d..c9e089fa9f 100644
--- a/src/crypto/tls/tls_cf.go
+++ b/src/crypto/tls/tls_cf.go
@@ -5,9 +5,9 @@ package tls
 
 import (
+	"time"
+
 	circlPki "github.com/cloudflare/circl/pki"
 	circlSign "github.com/cloudflare/circl/sign"
-
 	"github.com/cloudflare/circl/sign/eddilithium3"
-	"time"
 )
 
diff --git a/src/crypto/tls/tls_cf_circl_test.go b/src/crypto/tls/tls_cf_circl_test.go
index 92c577852e..e731e8b2f6 100644
--- a/src/crypto/tls/tls_cf_circl_test.go
+++ b/src/crypto/tls/tls_cf_circl_test.go
@@ -9,9 +9,10 @@ import (
 	"crypto/x509/pkix"
 	"fmt"
-	"github.com/cloudflare/circl/sign"
-	"github.com/cloudflare/circl/sign/eddilithium3"
 	"math/big"
 	"testing"
 	"time"
+
+	"github.com/cloudflare/circl/sign"
+	"github.com/cloudflare/circl/sign/eddilithium3"
 )
 
diff --git a/src/crypto/x509/parser.go b/src/crypto/x509/parser.go
index 00fb32828b..7c8117d616 100644
--- a/src/crypto/x509/parser.go
+++ b/src/crypto/x509/parser.go
@@ -26,8 +26,8 @@ import (
 	"unicode/utf8"
 
+	circlPki "github.com/cloudflare/circl/pki"
+
 	"golang.org/x/crypto/cryptobyte"
 	cryptobyte_asn1 "golang.org/x/crypto/cryptobyte/asn1"
-
-	circlPki "github.com/cloudflare/circl/pki"
 )
 
diff --git a/src/crypto/x509/x509_cf.go b/src/crypto/x509/x509_cf.go
index b597ad78f4..aab1c6480c 100644
--- a/src/crypto/x509/x509_cf.go
+++ b/src/crypto/x509/x509_cf.go
@@ -2,11 +2,10 @@ package x509
 
 import (
+	"crypto"
+	"encoding/asn1"
+
 	circlPki "github.com/cloudflare/circl/pki"
 	circlSign "github.com/cloudflare/circl/sign"
-
 	"github.com/cloudflare/circl/sign/eddilithium3"
-
-	"crypto"
-	"encoding/asn1"
 )
 
```
</details>

**Comment by Lekensteyn** [comment]
Changes since last push:

- Fixed imports ordering style, see comment https://github.com/cloudflare/go/pull/147#discussion_r1190189292 for the diff
- Moved the CFEventHandler changes to the original commits that introduced the changes and updated commit messages. See below for the diff.

Once approved, it can be force-pushed to the `cf` branch.

<details>
<summary>git range-diff 5e7998ded1...1fafe23f11</summary>

```
1:  a9e98095e1 ! 1:  40b3758c12 crypto/tls: add Config.CFEventHandler to expose intra-handshake timing
    @@ Metadata
     Author: Christopher Wood <chriswood@cloudflare.com>
     
      ## Commit message ##
    -    crypto/tls: add Config.CFEventHandler to expose intra-handshake timing
    +    crypto/tls: expose inter-handshake timing via CFEventHandlerContextKey
     
         Add basic support for handshake metrics:
     
    -    * Adds a `CFEventHandler()` callback to `Config` which can be called at
    -      various points during the handshake to respond to various events.
    +    * Adds the ability to set a callback via the CFEventHandlerContextKey
    +      context value on the handshake context. It will be called at various
    +      points during the handshake to respond to various events. See #146.
         * Use this callback to expose client and server intra-handshake state
           machine durations, respectively. Each event records elapsed timestamps
           (durations) for relevant events during the course of a connection,
    @@ Commit message
           This will be useful for recording intra-stack costs of TLS extensions
           such as ECH and KEMTLS.
     
    - ## src/crypto/tls/common.go ##
    -@@ src/crypto/tls/common.go: type Config struct {
    - 	// used for debugging.
    - 	KeyLogWriter io.Writer
    - 
    -+	// CFEventHandler, if set, is called by the client and server at various
    -+	// points during the handshake to handle specific events. This is used
    -+	// primarily for collecting metrics.
    -+	//
    -+	// NOTE: This feature is used to implement Cloudflare-internal features.
    -+	// This feature is unstable and applications MUST NOT depend on it.
    -+	CFEventHandler func(event CFEvent)
    -+
    - 	// mutex protects sessionTicketKeys and autoSessionTicketKeys.
    - 	mutex sync.RWMutex
    - 	// sessionTicketKeys contains zero or more ticket keys. If set, it means
    -@@ src/crypto/tls/common.go: func (c *Config) Clone() *Config {
    - 		DynamicRecordSizingDisabled: c.DynamicRecordSizingDisabled,
    - 		Renegotiation:               c.Renegotiation,
    - 		KeyLogWriter:                c.KeyLogWriter,
    -+		CFEventHandler:              c.CFEventHandler,
    - 		sessionTicketKeys:           c.sessionTicketKeys,
    - 		autoSessionTicketKeys:       c.autoSessionTicketKeys,
    - 	}
    +     [pwu: Go 1.20.4: moved Config.CFEventHandler to context value]
     
      ## src/crypto/tls/conn.go ##
    +@@ src/crypto/tls/conn.go: type Conn struct {
    + 	activeCall atomic.Int32
    + 
    + 	tmp [16]byte
    ++
    ++	// cfEventHandler is called at several points during the handshake if
    ++	// set. See also CFEventHandlerContextKey.
    ++	cfEventHandler func(event CFEvent)
    + }
    + 
    + // Access to net.Conn methods.
    +@@ src/crypto/tls/conn.go: func (c *Conn) handshakeContext(ctx context.Context) (ret error) {
    + 		return nil
    + 	}
    + 
    ++	if handler, ok := handshakeCtx.Value(CFEventHandlerContextKey{}).(func(event CFEvent)); ok {
    ++		c.cfEventHandler = handler
    ++	}
    ++
    + 	c.in.Lock()
    + 	defer c.in.Unlock()
    + 
     @@ src/crypto/tls/conn.go: func (c *Conn) VerifyHostname(host string) error {
      	}
      	return c.peerCertificates[0].VerifyHostname(host)
      }
     +
    ++// CFEventHandlerContextKey is a context key. It can be set on a TLS client or
    ++// server before the handshake has started. The associated value must be of type
    ++// func(event tls.CFEvent). This function will be called at various points
    ++// during the handshake for collecting metrics.
    ++//
    ++// NOTE: it can also be called at the end of the connection for certain ECH
    ++// metrics. This might be removed in the future.
    ++//
    ++// NOTE: This feature is used to implement Cloudflare-internal features.
    ++// This feature is unstable and applications MUST NOT depend on it.
    ++type CFEventHandlerContextKey struct{}
    ++
     +func (c *Conn) handleCFEvent(event CFEvent) {
    -+	if c.config.CFEventHandler != nil {
    -+		c.config.CFEventHandler(event)
    ++	if c.cfEventHandler != nil {
    ++		c.cfEventHandler(event)
     +	}
     +}
     
    @@ src/crypto/tls/metrics_test.go (new)
     +package tls
     +
     +import (
    ++	"context"
     +	"crypto/x509"
     +	"fmt"
     +	"io"
    @@ src/crypto/tls/metrics_test.go (new)
     +	c, s := localPipe(t)
     +	errChan := make(chan error)
     +
    -+	clientConfig.CFEventHandler = timingState.eventHandler
    -+	serverConfig.CFEventHandler = timingState.eventHandler
    -+
     +	go func() {
     +		cli := Client(c, clientConfig)
    -+		err := cli.Handshake()
    ++		cCtx := context.WithValue(context.Background(), CFEventHandlerContextKey{}, timingState.eventHandler)
    ++		err := cli.HandshakeContext(cCtx)
     +		if err != nil {
     +			errChan <- fmt.Errorf("client: %v", err)
     +			c.Close()
    @@ src/crypto/tls/metrics_test.go (new)
     +	}()
     +
     +	server := Server(s, serverConfig)
    -+	err = server.Handshake()
    ++	sCtx := context.WithValue(context.Background(), CFEventHandlerContextKey{}, timingState.eventHandler)
    ++	err = server.HandshakeContext(sCtx)
     +	if err == nil {
     +		if _, err := io.WriteString(server, sentinel); err != nil {
     +			t.Errorf("failed to call server.Write: %v", err)
    @@ src/crypto/tls/tls_cf.go: func init() {
     +		start: timer(),
     +	}
     +}
    -
    - ## src/crypto/tls/tls_test.go ##
    -@@ src/crypto/tls/tls_test.go: func TestCloneNonFuncFields(t *testing.T) {
    - 		switch fn := typ.Field(i).Name; fn {
    - 		case "Rand":
    - 			f.Set(reflect.ValueOf(io.Reader(os.Stdin)))
    --		case "Time", "GetCertificate", "GetConfigForClient", "VerifyPeerCertificate", "VerifyConnection", "GetClientCertificate":
    -+		case "Time", "GetCertificate", "GetConfigForClient", "VerifyPeerCertificate", "VerifyConnection", "GetClientCertificate", "CFEventHandler":
    - 			// DeepEqual can't compare functions. If you add a
    - 			// function field to this list, you must also change
    - 			// TestCloneFuncFields to ensure that the func field is
2:  2a51f15b29 ! 2:  c8f4a7e051 crypto/tls: implement draft-ietf-tls-subcerts-10
    @@ src/crypto/tls/common.go: type CertificateRequestInfo struct {
      	Version uint16
      
     @@ src/crypto/tls/common.go: type Config struct {
    - 	// This feature is unstable and applications MUST NOT depend on it.
    - 	CFEventHandler func(event CFEvent)
    + 	// used for debugging.
    + 	KeyLogWriter io.Writer
      
     +	// SupportDelegatedCredential is true if the client or server is willing
     +	// to negotiate the delegated credential extension.
    @@ src/crypto/tls/common.go: func (c *Config) Clone() *Config {
      		Renegotiation:               c.Renegotiation,
      		KeyLogWriter:                c.KeyLogWriter,
     +		SupportDelegatedCredential:  c.SupportDelegatedCredential,
    - 		CFEventHandler:              c.CFEventHandler,
      		sessionTicketKeys:           c.sessionTicketKeys,
      		autoSessionTicketKeys:       c.autoSessionTicketKeys,
    + 	}
     @@ src/crypto/tls/common.go: func (c *Config) writeKeyLog(label string, clientRandom, secret []byte) error {
      // and is only for debugging, so a global mutex saves space.
      var writerMutex sync.Mutex
3:  23217410ed ! 3:  3c7a87eb55 crypto/tls: implement draft-ietf-tls-esni-13
    @@ src/crypto/tls/common.go: type Config struct {
     +	// payload.
     +	ServerECHProvider ECHProvider
     +
    - 	// CFEventHandler, if set, is called by the client and server at various
    - 	// points during the handshake to handle specific events. This is used
    - 	// primarily for collecting metrics.
    + 	// SupportDelegatedCredential is true if the client or server is willing
    + 	// to negotiate the delegated credential extension.
    + 	// This can only be used with TLS 1.3.
     @@ src/crypto/tls/common.go: func (c *Config) Clone() *Config {
      		Renegotiation:               c.Renegotiation,
      		KeyLogWriter:                c.KeyLogWriter,
    @@ src/crypto/tls/common.go: func (c *Config) Clone() *Config {
     +		ECHEnabled:                  c.ECHEnabled,
     +		ClientECHConfigs:            c.ClientECHConfigs,
     +		ServerECHProvider:           c.ServerECHProvider,
    - 		CFEventHandler:              c.CFEventHandler,
      		sessionTicketKeys:           c.sessionTicketKeys,
      		autoSessionTicketKeys:       c.autoSessionTicketKeys,
    + 	}
     @@ src/crypto/tls/common.go: func (c *Config) supportedVersions(isClient bool) []uint16 {
      	return versions
      }
    @@ src/crypto/tls/conn.go: import (
      
      // A Conn represents a secured connection.
     @@ src/crypto/tls/conn.go: type Conn struct {
    - 	activeCall atomic.Int32
    - 
    - 	tmp [16]byte
    + 	// cfEventHandler is called at several points during the handshake if
    + 	// set. See also CFEventHandlerContextKey.
    + 	cfEventHandler func(event CFEvent)
     +
     +	// State used for the ECH extension.
     +	ech struct {
    @@ src/crypto/tls/ech_test.go (new)
     +
     +import (
     +	"bytes"
    ++	"context"
     +	"crypto/rand"
     +	"crypto/x509"
     +	"encoding/pem"
    @@ src/crypto/tls/ech_test.go (new)
     +	serverCh := make(chan echTestResult, 1)
     +	go func() {
     +		var res echTestResult
    -+		serverConfig.CFEventHandler = res.eventHandler
     +		serverConn, err := ln.Accept()
     +		if err != nil {
     +			res.err = err
    @@ src/crypto/tls/ech_test.go (new)
     +			serverCh <- res
     +		}()
     +
    -+		if err := server.Handshake(); err != nil {
    ++		sCtx := context.WithValue(context.Background(), CFEventHandlerContextKey{}, res.eventHandler)
    ++		if err := server.HandshakeContext(sCtx); err != nil {
     +			res.err = err
     +			return
     +		}
    @@ src/crypto/tls/ech_test.go (new)
     +		res.connState = server.ConnectionState()
     +	}()
     +
    -+	clientConfig.CFEventHandler = clientRes.eventHandler
    -+	client, err := Dial("tcp", ln.Addr().String(), clientConfig)
    ++	cCtx := context.WithValue(context.Background(), CFEventHandlerContextKey{}, clientRes.eventHandler)
    ++	clientDialer := &Dialer{Config: clientConfig}
    ++	clientConn, err := clientDialer.DialContext(cCtx, "tcp", ln.Addr().String())
     +	if err != nil {
     +		serverRes = <-serverCh
     +		clientRes.err = err
     +		return
     +	}
    ++	client := clientConn.(*Conn)
     +	defer client.Close()
     +
     +	_, err = client.Write(testMessage)
    @@ src/crypto/tls/tls_test.go: func TestCloneNonFuncFields(t *testing.T) {
      		switch fn := typ.Field(i).Name; fn {
      		case "Rand":
      			f.Set(reflect.ValueOf(io.Reader(os.Stdin)))
    --		case "Time", "GetCertificate", "GetConfigForClient", "VerifyPeerCertificate", "VerifyConnection", "GetClientCertificate", "CFEventHandler":
    -+		case "Time", "GetCertificate", "GetConfigForClient", "VerifyPeerCertificate", "VerifyConnection", "GetClientCertificate", "ServerECHProvider", "CFEventHandler":
    +-		case "Time", "GetCertificate", "GetConfigForClient", "VerifyPeerCertificate", "VerifyConnection", "GetClientCertificate":
    ++		case "Time", "GetCertificate", "GetConfigForClient", "VerifyPeerCertificate", "VerifyConnection", "GetClientCertificate", "ServerECHProvider":
      			// DeepEqual can't compare functions. If you add a
      			// function field to this list, you must also change
      			// TestCloneFuncFields to ensure that the func field is
4:  30b290ea92 ! 4:  1fafe23f11 Add hybrid post-quantum key agreement.
    @@ Metadata
     Author: Bas Westerbaan <bas@cloudflare.com>
     
      ## Commit message ##
    -    Add hybrid post-quantum key agreement.
    +    crypto/tls: Add hybrid post-quantum key agreement
     
    -    Adds X25519Kyber512Draft00 and X25519Kyber768Draft00 hybrid post-quantum key
    -    agreements with temporary group identifiers.
    +    Adds X25519Kyber512Draft00, X25519Kyber768Draft00, and
    +    P256Kyber768Draft00 hybrid post-quantum key agreements with temporary
    +    group identifiers.
     
         The hybrid post-quantum key exchanges uses plain X{25519,448} instead
         of HPKE, which we assume will be more likely to be adopted. The order
    @@ src/crypto/tls/cfkem_test.go (new)
     +package tls
     +
     +import (
    ++	"context"
     +	"fmt"
     +	"testing"
     +
    @@ src/crypto/tls/cfkem_test.go (new)
     +			X25519,
     +		}
     +	}
    -+	clientConfig.CFEventHandler = func(ev CFEvent) {
    ++	clientCFEventHandler := func(ev CFEvent) {
     +		switch e := ev.(type) {
     +		case CFEventTLSNegotiatedNamedKEX:
     +			clientSelectedKEX = &e.KEX
    @@ src/crypto/tls/cfkem_test.go (new)
     +	}()
     +
     +	cli := Client(c, clientConfig)
    -+	clientErr := cli.Handshake()
    ++	cCtx := context.WithValue(context.Background(), CFEventHandlerContextKey{}, clientCFEventHandler)
    ++	clientErr := cli.HandshakeContext(cCtx)
     +	serverErr := <-done
     +	if clientErr != nil {
     +		t.Errorf("client error: %s", clientErr)
5:  5e7998ded1 < -:  ---------- crypto/tls: replace CFEventHandler by a context key

```
</details>

**Comment by Lekensteyn** [comment]
Changes since last push:
* Moved CFEvent code to a new file (cfevent.go) to make cherry-picking easier.
* Revert accidental changes to `src/vendor/golang.org/x/net/http2/hpack/hpack.go`. This issue was introduced with the Go 1.20.1 vendor rebase in #144. This was caused by a mismatch in go.mod and the actual go.mod contents as part of upstream commit 8e02cffd8e8a that got fixed by upstream commit 828b05cc647e as part of Go 1.20.2.
* Reordered commits such that the hybrid PQ key agreement patch and the related CFEvent patches come before ECH, DC, and CIRCL signature schemes patches (see filtered range diff below).
* Reworded some commit messages (see filtered range diff below).

```
$ git diff --stat 1fafe23f11 8284e397f7
 src/crypto/tls/cfevent.go                        | 184 +++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++
 src/crypto/tls/tls_cf.go                         | 180 ---------------------------------------------------------------------------------------------------------------------------------------------------
 src/vendor/golang.org/x/net/http2/hpack/hpack.go |  79 ++++++++++++++++++++++++++++++++++++++++-------------------------
 3 files changed, 233 insertions(+), 210 deletions(-)
```

```
$ git range-diff go1.20.4..1fafe23f11 go1.20.4..8284e397f7 | awk '/^    . ## /||/^    @@ [^C]/{skip=1} !/^    /{skip=0}; !skip'
 1:  75a9a74aff =  1:  75a9a74aff VERSION: add cf tag
 2:  07c345757a =  2:  07c345757a cmd/go/internal/imports, go/build: add cfgo build tag
 3:  79dbb4eb9b =  3:  79dbb4eb9b .github: add docker-compose.yaml and GitHub Action for CI
 4:  38dd0ba444 =  4:  38dd0ba444 .github: replace upstream templates and CODE_OF_CONDUCT.md
 5:  7766b6542e =  5:  7766b6542e cmd/api: ignore CF-specific API changes
 6:  61985f6c98 =  6:  61985f6c98 go/build: don't check dependencies of vendored circl library
 8:  40b3758c12 !  7:  96832dcf5b crypto/tls: expose inter-handshake timing via CFEventHandlerContextKey
    @@ Commit message
           such as ECH and KEMTLS.

          [pwu: Go 1.20.4: moved Config.CFEventHandler to context value]
    +     [pwu: Go 1.20.4: moved CFEvent code from tls_cf.go to cfevent.go]
    +
11:  1fafe23f11 !  8:  dd708185f7 crypto/tls: Add hybrid post-quantum key agreement
    @@ Commit message
         Adds CFEvents to detect `HelloRetryRequest`s and to signal which
         key agreement was used.

    -    Cf #121 #122 #123 #132
    -
         Co-authored-by: Christopher Wood <caw@heapingbits.net>

    -    [ bas, 1.20.1: also adds P256Kyber768Draft00 ]
    +     [bas, 1.20.1: also adds P256Kyber768Draft00]
    +     [pwu, 1.20.4: updated circl to v1.3.3, moved code to cfevent.go]
    +
 7:  d0a3c75ee0 !  9:  743259c6b3 crypto: add support for CIRCL signature schemes
 9:  c8f4a7e051 ! 10:  130a40e5e8 crypto/tls: implement draft-ietf-tls-subcerts-10
    @@ Commit message

         * Define API for delegated credentials so they are fetched using the
           same mechanisms used to fetch certificates
    -    * Allow the usage of other keyUsage when checking for the dc extension.
    -
    -    Fixes issues in earlier patch, addressing #127, #128, #129, #130, and #131.
    -    Add tool for generating delegated credentials.
    +    * Allow the usage of other keyUsage when checking for the DC extension.
    +    * Add tool for generating delegated credentials.

         Co-authored-by: Jonathan Hoyland <jhoyland@cloudflare.com>

10:  3c7a87eb55 ! 11:  8284e397f7 crypto/tls: implement draft-ietf-tls-esni-13
```

**Pull request merged**
